### ISSUE TITLE
GC failure in L1 block reader

### ISSUE THREAD

**Issue by dhammika** [neondatabase/neon#1594]
Get page lookup failures with low churn threshold. I was trying a different approach for #1484,  
```diff
--- a/pageserver/src/layered_repository.rs
+++ b/pageserver/src/layered_repository.rs
@@ -1824,7 +1824,7 @@ impl LayeredTimeline {
             // 2. Create new image layers for partitions that have been modified
             // "enough".
             for part in partitioning.parts.iter() {
-                if self.time_for_new_image_layer(part, lsn, 3)? {
+                if self.time_for_new_image_layer(part, lsn, 1)? {
                     self.create_image_layer(part, lsn)?;
                 }
             }

```

Test output,
```
FAILED test_runner/batch_others/test_branch_and_gc.py::test_branch_and_gc - asyncpg.exceptions._base.UnknownPostgresError: failed to read value from data file 000000067F000032A...
```

```
2022-04-29T04:02:16.266851Z ERROR query handler for 'do_gc f18eca09cb4744d5a7bda1fc913f7a44 3afc7c5b06ae59a4840b67dcae305572 0' failed: failed to read value from data file 000000067F000032AE000040000000000200-00000006800000000000000A890100000000__0000000002196BF8 at offset 2065388

Caused by:
    failed to fill whole buffer

Stack backtrace:
   0: anyhow::context::<impl anyhow::Context<T,E> for core::result::Result<T,E>>::with_context::{{closure}}
             at /home/d/.cargo/registry/src/github.com-1ecc6299db9ec823/anyhow-1.0.53/src/context.rs:58:30
   1: core::result::Result<T,E>::map_err
             at /rustc/db9d1b20bba1968c1ec1fc49616d4742c1725b4b/library/core/src/result.rs:853:27
   2: anyhow::context::<impl anyhow::Context<T,E> for core::result::Result<T,E>>::with_context
             at /home/d/.cargo/registry/src/github.com-1ecc6299db9ec823/anyhow-1.0.53/src/context.rs:58:9
   3: <pageserver::layered_repository::image_layer::ImageLayer as pageserver::layered_repository::storage_layer::Layer>::get_value_reconstruct_data
             at /home/d/src/github.com/zenithdb/zenith/pageserver/src/layered_repository/image_layer.rs:164:24
   4: pageserver::layered_repository::LayeredTimeline::get_reconstruct_data
             at /home/d/src/github.com/zenithdb/zenith/pageserver/src/layered_repository.rs:1462:26
   5: <pageserver::layered_repository::LayeredTimeline as pageserver::repository::Timeline>::get
             at /home/d/src/github.com/zenithdb/zenith/pageserver/src/layered_repository.rs:1072:9
   6: pageserver::layered_repository::LayeredTimeline::create_image_layer
             at /home/d/src/github.com/zenithdb/zenith/pageserver/src/layered_repository.rs:1887:27
   7: pageserver::layered_repository::LayeredTimeline::compact
             at /home/d/src/github.com/zenithdb/zenith/pageserver/src/layered_repository.rs:1829:21
   8: <pageserver::layered_repository::LayeredTimeline as pageserver::repository::Timeline>::checkpoint
             at /home/d/src/github.com/zenithdb/zenith/pageserver/src/layered_repository.rs:1090:17
   9: pageserver::layered_repository::LayeredRepository::gc_iteration_internal
```

Branch point `0/2192B20` is well outside lookup LSN `2196BF8`
```
2022-04-29T04:01:33.647781Z  INFO /timeline_create{tenant=f18eca09cb4744d5a7bda1fc913f7a44 new_timeline=None lsn=Some(0/2192B20)}: branched timeline cb02935f0cc43ca77bf885612a571c96 from 3afc7c5b06ae59a4840b67dcae305572 at 0/2192B20
```

**Comment by knizhnik** [comment]
Is there any hint how to reproduce the problem?
I used test_branch_and_gc.py test from #1484 and add 'image_creation_threshold': '1' to config.
I have repeated this test 100 times with release build of main branch but failed to reproduce the problem.
@dhammika are you still able to reproduce it?'

**Comment by dhammika** [comment]
> Is there any hint how to reproduce the problem? I used test_branch_and_gc.py test from #1484 and add 'image_creation_threshold': '1' to config. I have repeated this test 100 times with release build of main branch but failed to reproduce the problem. @dhammika are you still able to reproduce it?'

Don't have any additional repro steps. I'll also check if I can repro it in my setup.

**Comment by knizhnik** [comment]
I still not able to reproduce the problem with this test. But I wonder why do we need to simulate async GC in the test, if we actually have GC thread in pageserver? We can just adjust GC parameters (make it happen more frequently) rather then manually initiate GC in the test. Manual  invocation of GC is needed only when we want to perform GC at some particular point.

**Comment by dhammika** [comment]
> I still not able to reproduce the problem with this test. But I wonder why do we need to simulate async GC in the test, if we actually have GC thread in pageserver? We can just adjust GC parameters (make it happen more frequently) rather then manually initiate GC in the test. Manual invocation of GC is needed only when we want to perform GC at some particular point.

Idea was to instrument everything from client side but changing GC settings might work too. Not sure why you don't see it in your setup. I see it failing in CI builds as well.

**Comment by knizhnik** [comment]
Fixed by commit [d301b8364]
